#VRML_SIM R2023b utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/floors/protos/RectangleArena.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/robots/robotis/turtlebot/protos/TurtleBot3Burger.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/tables/protos/Table.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/chairs/protos/WoodenChair.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/living_room_furniture/protos/Sofa.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/kitchen/fridge/protos/Fridge.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/humans/pedestrian/protos/Pedestrian.proto"

WorldInfo {
  title "SemanticVLN Final"
  basicTimeStep 32
}

Viewpoint {
  orientation -0.3 0.3 0.9 1.6
  position 0 -16 16
  follow "turtlebot3"
}

TexturedBackground { texture "empty_office" }
TexturedBackgroundLight { }

# 12x12 Open Arena
RectangleArena {
  floorSize 12 12
  floorTileSize 1 1
  wallHeight 2.5
  floorAppearance PBRAppearance { baseColor 0.85 0.85 0.82 roughness 0.4 }
  wallAppearance PBRAppearance { baseColor 0.92 0.9 0.88 roughness 0.5 }
}

# Robot at center
TurtleBot3Burger {
  translation 0 0 0
  rotation 0 0 1 0
  name "turtlebot3"
  controller "semantic_vln_controller"
  synchronization FALSE
  extensionSlot [
    GPS { }
    InertialUnit { }
    Lidar {
      name "lidar"
      translation 0 0 0.1
      numberOfLayers 1
      horizontalResolution 360
      fieldOfView 6.28
      minRange 0.12
      maxRange 5.0
    }
    Camera {
      translation 0 0 0.1
      name "camera"
      fieldOfView 1.047
      width 640
      height 480
    }
    RangeFinder {
      translation 0 0 0.1
      name "range-finder"
      fieldOfView 1.047
      width 640
      height 480
      minRange 0.1
      maxRange 8
    }
    Display { name "detection_display" width 640 height 480 }
    Display { name "slam_display" width 200 height 200 }
  ]
}

# ===== KITCHEN (Top-Right) - SPREAD OUT =====
Fridge {
  translation 5 5 0
  rotation 0 0 1 -2.36
  name "kitchen_fridge"
}
Table {
  translation 2 3 0
  name "kitchen_table"
  size 0.9 0.6 0.72
}
WoodenChair {
  translation 1 3 0
  rotation 0 0 1 0
  name "kitchen_chair"
}

# ===== LIVING ROOM (Bottom-Left) =====
Sofa {
  translation -5 -4 0
  rotation 0 0 1 1.57
  name "living_sofa"
}
Table {
  translation -3 -4.5 0
  name "living_table"
  size 0.7 0.5 0.4
}

# ===== BEDROOM (Top-Left) =====
Solid {
  translation -5 5 0.2
  children [
    Shape {
      appearance PBRAppearance { baseColor 0.5 0.35 0.25 roughness 0.6 }
      geometry Box { size 1.6 1.0 0.35 }
    }
  ]
  name "bedroom_bed"
}

# ===== BATHROOM (Bottom-Right) =====
Solid {
  translation 5 -5 0.4
  children [
    Shape {
      appearance PBRAppearance { baseColor 0.95 0.95 0.95 roughness 0.1 }
      geometry Box { size 0.5 0.4 0.7 }
    }
  ]
  name "bathroom_sink"
}

# ===== PERSON (walks in OPEN area away from furniture) =====
Pedestrian {
  translation -2 0 1.27
  name "person_1"
  controller "pedestrian"
  controllerArgs [
    "--speed=0.2"
    "--trajectory=-2 0, -2 2, 0 2, 0 0, -2 0"
  ]
}
